[NATS] Program crashes on invalid server URL

**Description:**
$title.

**Steps to reproduce:**
```ballerina
import ballerina/io;
import ballerina/nats;
 
nats:Connection conn = new("/localhost:4222"); // invalid
 
listener nats:Listener lis = new(conn);
 
@nats:SubscriptionConfig {
    subject: "demo"
}
service demo on lis {
    resource function onMessage(nats:Message msg, string data) {
        io:println(data);
    }
 
    resource function onError(nats:Message msg, nats:Error err) {
        io:println(err);
    }
}
```

ballerina-internal.log
```
[2019-11-04 08:25:49,823] SEVERE {b7a.log.crash} - Bad server URL: /localhost:4222 
java.lang.IllegalArgumentException: Bad server URL: /localhost:4222
	at io.nats.client.Options$Builder.servers(Options.java:717)
	at org.ballerinalang.nats.connection.Init.init(Init.java:103)
	at ballerina.nats.$value$Connection.init(connection/connection.bal:33)
	at ballerina.nats.$value$Connection.call(connection/connection.bal)
	at ballerina.nats.$value$Connection.__init(connection/connection.bal:30)
	at ballerina.nats.$value$Connection.call(connection/connection.bal)
	at ___init.___init_(.:4)
	at ___init.$moduleInit(.:65535)
	at ___init.$lambda$$moduleInit$(.)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:426)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:218)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:191)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.URISyntaxException: unable to parse server URI: /localhost:4222
	at io.nats.client.Options.parseURIForServer(Options.java:1509)
	at io.nats.client.Options$Builder.servers(Options.java:715)
	... 12 more
 
[2019-11-04 08:25:49,828] SEVERE {b7a.log.crash} - Bad server URL: /localhost:4222 
java.lang.IllegalArgumentException: Bad server URL: /localhost:4222
	at io.nats.client.Options$Builder.servers(Options.java:717)
	at org.ballerinalang.nats.connection.Init.init(Init.java:103)
	at ballerina.nats.$value$Connection.init(connection/connection.bal:33)
	at ballerina.nats.$value$Connection.call(connection/connection.bal)
	at ballerina.nats.$value$Connection.__init(connection/connection.bal:30)
	at ballerina.nats.$value$Connection.call(connection/connection.bal)
	at ___init.___init_(.:4)
	at ___init.$moduleInit(.:65535)
	at ___init.$lambda$$moduleInit$(.)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:426)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:218)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:191)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.URISyntaxException: unable to parse server URI: /localhost:4222
	at io.nats.client.Options.parseURIForServer(Options.java:1509)
	at io.nats.client.Options$Builder.servers(Options.java:715)
	... 12 more
 
[2019-11-04 08:25:49,836] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
	at ___init.___stop_(.:6)
	at ___init.$lambda$___stop_(.)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:426)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:218)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:191)
	at org.ballerinalang.jvm.scheduling.Scheduler.start(Scheduler.java:178)
	at ___init.$moduleStop(.)
	at ___init$SignalListener.run(Unknown Source)
 
[2019-11-04 08:25:49,840] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
	at ___init.___stop_(.:6)
	at ___init.$lambda$___stop_(.)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:426)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:218)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:191)
	at org.ballerinalang.jvm.scheduling.Scheduler.start(Scheduler.java:178)
	at ___init.$moduleStop(.)
	at ___init$SignalListener.run(Unknown Source)
```

**Affected Versions:**
1.0.3

--- a/stdlib/http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/Respond.java
+++ b/stdlib/http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/Respond.java
@@ -66,7 +66,10 @@ public class Respond extends ConnectionAction {
             String errorMessage = "Couldn't complete the respond operation as the response has been already used.";
             HttpUtil.sendOutboundResponse(inboundRequestMsg, HttpUtil.createErrorMessage(errorMessage, 500));
             unBlockStrand(strand);
-            log.debug(errorMessage);
+            if (log.isDebugEnabled()) {
+                log.debug("Couldn't complete the respond operation for the sequence id of the request: {} " +
+                        "as the response has been already used.", inboundRequestMsg.getSequenceId());
+            }
             return HttpUtil.createHttpError(errorMessage, HttpErrorType.GENERIC_LISTENER_ERROR);
         }
         outboundResponseObj.addNativeData(HttpConstants.DIRTY_RESPONSE, true);

--- a/stdlib/messaging/nats/src/main/java/org/ballerinalang/nats/connection/Init.java
+++ b/stdlib/messaging/nats/src/main/java/org/ballerinalang/nats/connection/Init.java
@@ -76,51 +76,51 @@ public class Init {
         } else {
             serverUrls = new String[]{urlString};
         }
-        opts.servers(serverUrls);
-
-        // Add connection name.
-        opts.connectionName(connectionConfig.getStringValue(CONNECTION_NAME));
-
-        // Add max reconnect.
-        opts.maxReconnects(Math.toIntExact(connectionConfig.getIntValue(MAX_RECONNECT)));
-
-        // Add reconnect wait.
-        opts.reconnectWait(Duration.ofSeconds(connectionConfig.getIntValue(RECONNECT_WAIT)));
-
-        // Add connection timeout.
-        opts.connectionTimeout(Duration.ofSeconds(connectionConfig.getIntValue(CONNECTION_TIMEOUT)));
-
-        // Add ping interval.
-        opts.pingInterval(Duration.ofMinutes(connectionConfig.getIntValue(PING_INTERVAL)));
-
-        // Add max ping out.
-        opts.maxPingsOut(Math.toIntExact(connectionConfig.getIntValue(MAX_PINGS_OUT)));
-
-        // Add inbox prefix.
-        opts.inboxPrefix(connectionConfig.getStringValue(INBOX_PREFIX));
-
-        List<ObjectValue> serviceList = Collections.synchronizedList(new ArrayList<>());
-        // Add NATS connection listener.
-        opts.connectionListener(new DefaultConnectionListener());
-
-        // Add NATS error listener.
-        if (connectionConfig.getBooleanValue(ENABLE_ERROR_LISTENER)) {
-            ErrorListener errorListener = new DefaultErrorListener();
-            opts.errorListener(errorListener);
-        }
-
-        // Add noEcho.
-        if (connectionConfig.getBooleanValue(NO_ECHO)) {
-            opts.noEcho();
-        }
-
-        MapValue secureSocket = connectionConfig.getMapValue(Constants.CONNECTION_CONFIG_SECURE_SOCKET);
-        if (secureSocket != null) {
-            SSLContext sslContext = getSSLContext(secureSocket);
-            opts.sslContext(sslContext);
-        }
-
         try {
+            opts.servers(serverUrls);
+
+            // Add connection name.
+            opts.connectionName(connectionConfig.getStringValue(CONNECTION_NAME));
+
+            // Add max reconnect.
+            opts.maxReconnects(Math.toIntExact(connectionConfig.getIntValue(MAX_RECONNECT)));
+
+            // Add reconnect wait.
+            opts.reconnectWait(Duration.ofSeconds(connectionConfig.getIntValue(RECONNECT_WAIT)));
+
+            // Add connection timeout.
+            opts.connectionTimeout(Duration.ofSeconds(connectionConfig.getIntValue(CONNECTION_TIMEOUT)));
+
+            // Add ping interval.
+            opts.pingInterval(Duration.ofMinutes(connectionConfig.getIntValue(PING_INTERVAL)));
+
+            // Add max ping out.
+            opts.maxPingsOut(Math.toIntExact(connectionConfig.getIntValue(MAX_PINGS_OUT)));
+
+            // Add inbox prefix.
+            opts.inboxPrefix(connectionConfig.getStringValue(INBOX_PREFIX));
+
+            List<ObjectValue> serviceList = Collections.synchronizedList(new ArrayList<>());
+            // Add NATS connection listener.
+            opts.connectionListener(new DefaultConnectionListener());
+
+            // Add NATS error listener.
+            if (connectionConfig.getBooleanValue(ENABLE_ERROR_LISTENER)) {
+                ErrorListener errorListener = new DefaultErrorListener();
+                opts.errorListener(errorListener);
+            }
+
+            // Add noEcho.
+            if (connectionConfig.getBooleanValue(NO_ECHO)) {
+                opts.noEcho();
+            }
+
+            MapValue secureSocket = connectionConfig.getMapValue(Constants.CONNECTION_CONFIG_SECURE_SOCKET);
+            if (secureSocket != null) {
+                SSLContext sslContext = getSSLContext(secureSocket);
+                opts.sslContext(sslContext);
+            }
+
             Connection natsConnection = Nats.connect(opts.build());
             connectionObject.addNativeData(Constants.NATS_CONNECTION, natsConnection);
             connectionObject.addNativeData(Constants.CONNECTED_CLIENTS, new AtomicInteger(0));
@@ -131,6 +131,10 @@ public class Init {
             String errorMsg = "Error while setting up a connection. " +
                     (e.getCause() != null ? e.getCause().getMessage() : e.getMessage());
             throw Utils.createNatsError(errorMsg);
+        } catch (IllegalStateException e) {
+            NatsMetricsUtil.reportError(NatsObservabilityConstants.CONTEXT_CONNECTION,
+                    NatsObservabilityConstants.ERROR_TYPE_CONNECTION);
+            throw Utils.createNatsError(e.getMessage());
         }
     }
 

--- a/stdlib/messaging/nats/src/main/java/org/ballerinalang/nats/connection/Init.java
+++ b/stdlib/messaging/nats/src/main/java/org/ballerinalang/nats/connection/Init.java
@@ -131,7 +131,7 @@ public class Init {
             String errorMsg = "Error while setting up a connection. " +
                     (e.getCause() != null ? e.getCause().getMessage() : e.getMessage());
             throw Utils.createNatsError(errorMsg);
-        } catch (IllegalStateException e) {
+        } catch (IllegalArgumentException e) {
             NatsMetricsUtil.reportError(NatsObservabilityConstants.CONTEXT_CONNECTION,
                     NatsObservabilityConstants.ERROR_TYPE_CONNECTION);
             throw Utils.createNatsError(e.getMessage());

--- a/stdlib/messaging/nats/src/test/java/org/ballerinalang/nats/streaming/NatsStreamingSubscriberCompilationTest.java
+++ b/stdlib/messaging/nats/src/test/java/org/ballerinalang/nats/streaming/NatsStreamingSubscriberCompilationTest.java
@@ -20,6 +20,7 @@ package org.ballerinalang.nats.streaming;
 import org.ballerinalang.test.util.BAssertUtil;
 import org.ballerinalang.test.util.BCompileUtil;
 import org.ballerinalang.test.util.CompileResult;
+import org.ballerinalang.util.exceptions.BLangRuntimeException;
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
@@ -114,6 +115,19 @@ public class NatsStreamingSubscriberCompilationTest {
                         + "second paramter (optional) type is "
                         + "byte[] | boolean | string | int | float | decimal | xml | json | record {}",
                 28, 6);
+    }
+
+    @Test(description = "Test invalid connection URL")
+    public void testInvalidURL() {
+        String resourceRoot = Paths.get("src", "test", "resources").toAbsolutePath().toString();
+        Path testResourceRoot = Paths.get(resourceRoot, "test-src", "basic");
+        try {
+            BCompileUtil.compileOffline(testResourceRoot.resolve("nats_invalid_url.bal").toString());
+        } catch (BLangRuntimeException e) {
+            String actualMsg = e.getMessage();
+            String expectedErrorMsg = "Bad server URL: /localhost:4222";
+            Assert.assertTrue(actualMsg.contains(expectedErrorMsg));
+        }
     }
 
     private CompileResult compileBasicTests(String fileName) {

--- a/stdlib/messaging/nats/src/test/resources/test-src/basic/nats_invalid_url.bal
+++ b/stdlib/messaging/nats/src/test/resources/test-src/basic/nats_invalid_url.bal
@@ -1,0 +1,35 @@
+// Copyright (c) 2020 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/io;
+import ballerina/nats;
+
+nats:Connection conn = new("/localhost:4222");
+
+listener nats:Listener lis = new(conn);
+
+@nats:SubscriptionConfig {
+    subject: "demo"
+}
+service demo on lis {
+    resource function onMessage(nats:Message msg, string data) {
+        io:println(data);
+    }
+
+    resource function onError(nats:Message msg, nats:Error err) {
+        io:println(err);
+    }
+}

